### PR TITLE
Make substrate generic

### PR DESCRIPTION
**Pull request by gavofyork**
With the eventual aim to remove `substrate/primitives/block.rs` entirely. This will mean making the `bft` modules generic since they have substantial dependencies on several of the types inherent in `Header`. Yet to do:

- [x] place `Hashing` type in `Header` in favour of Hash;
- [x] get runtime compiling;
- [x] move `substrate_primitives::bft` module into `substrate_runtime_primitives` and make generic over `substrate_runtime_primitives::{Block, Header, Hashing}` traits;
- [x] make `substrate/bft` similarly generic;
- [x] make `substrate/network` similarly generic;
- [x] utilise the new generic types for `polkadot` code's `Block`, `Header` &c;
- [x] port the `polkadot/consensus` module to use the new types.

### CODE REVIEW

File: substrate/rpc/src/chain/mod.rs [hunk 1]
```diff
@@ -40,10 +40,11 @@ build_rpc_trait! {
 	/// Polkadot blockchain API
 	pub trait ChainApi {
 		type Metadata;
+		ASSOCIATED type Header;
```
Comment (rphmeier) [inline substrate/rpc/src/chain/mod.rs]: why caps?

File: substrate/runtime/primitives/src/traits.rs [hunk 2]
```diff
@@ -147,10 +149,86 @@ impl<A: Executable, B: Executable> Executable for (A, B) {
 	}
 }
 
+/// Abstraction around hashing
+pub trait Hashing {
```
Comment (rphmeier) [inline substrate/runtime/primitives/src/traits.rs]: more idiomatic to call it `Hash` since it describes a hash function

File: substrate/runtime/primitives/src/traits.rs [hunk 3]
```diff
@@ -147,10 +149,86 @@ impl<A: Executable, B: Executable> Executable for (A, B) {
...
+	/// Produce the hash of some byte-slice.
+	fn hash(s: &[u8]) -> Self::Output;
+
+	/// Produce the hash of some codec-encodable value.
+	fn hash_of<S: codec::Slicable>(s: &S) -> Self::Output {
+		codec::Slicable::using_encoded(s, Self::hash)
+	}
+
+	/// Produce the patricia-trie root of a mapping from indices to byte slices.
+	fn enumerated_trie_root(items: &[&[u8]]) -> Self::Output;
```
Comment (rphmeier) [inline substrate/runtime/primitives/src/traits.rs]: this and other `trie_root` functions should be able to go away with upcoming generalization. Although it will use a slightly different `Hash` trait

File: substrate/runtime/primitives/src/traits.rs [hunk 4]
```diff
@@ -147,10 +149,86 @@ impl<A: Executable, B: Executable> Executable for (A, B) {
...
+
+	/// The Patricia tree root of the given mapping as an iterator.
+	fn trie_root<
+		I: IntoIterator<Item = (A, B)>,
+		A: AsRef<[u8]> + Ord,
+		B: AsRef<[u8]>
+	>(input: I) -> Self::Output;
+
+	/// Acquire the global storage root.
+	fn storage_root() -> Self::Output;
```
Comment (rphmeier) [inline substrate/runtime/primitives/src/traits.rs]: that doesn't seem like something that belongs in a hash trait, rather it should be some functionality which is parameterized by it

**Comment by rphmeier** [comment]
icing until #113 is in (some conflicting changes)

File: substrate/runtime/primitives/src/generic.rs [hunk 5]
```diff
@@ -353,6 +337,28 @@ impl<Number, Hash, DigestItem> traits::Header for Header<Number, Hash, DigestIte
...
+	/// Identify by block header hash.
+	Hash(Hash),
+	/// Identify by block number.
+	Number(Number),
+}
+
+#[cfg(feature = "std")]
+impl<Block: BlockT> fmt::Display for Id<Block> {
+	fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+		format!("{:?}", self)
```
Comment (rphmeier) [inline substrate/runtime/primitives/src/generic.rs]: should be `write!(f, "{:?}", self)`, currently it just creates a string on the heap and drops it.

File: substrate/rpc/src/chain/mod.rs (hunk 1 above)
Comment (gavofyork) [reply substrate/rpc/src/chain/mod.rs]: to distinguish from the rust-like grammar

File: substrate/runtime/primitives/src/traits.rs (hunk 2 above)
Comment (gavofyork) [reply substrate/runtime/primitives/src/traits.rs]: it's actually just moving existing code around, so rather out of scope for this pr

File: substrate/runtime/primitives/src/traits.rs (hunk 3 above)
Comment (gavofyork) [reply substrate/runtime/primitives/src/traits.rs]: agree. one for a later refactoring pr though.

File: substrate/runtime/primitives/src/traits.rs (hunk 4 above)
Comment (gavofyork) [reply substrate/runtime/primitives/src/traits.rs]: was already there and being used, so if there's any refactoring to be done, i would prefer to leave it for another pr.

File: substrate/rpc/src/chain/mod.rs (hunk 1 above)
Comment (rphmeier) [reply substrate/rpc/src/chain/mod.rs]: Ok, I might not understand how this works. What's really the difference between declaring `type Header` and `type Metadata`?

Comment (gavofyork) [reply substrate/rpc/src/chain/mod.rs]: `type Metadata;` already existed; it's some sort of marker that the macro uses to change the code that's generated.

i had to add the `ASSOCIATED` tag in order to allow for doc-comments of the type, since otherwise the grammar is ambiguous.

**Comment by dvdplm** [comment]
Trying to build this I get a compilation error like so:

```
error[E0271]: type mismatch resolving `<I as std::iter::IntoIterator>::Item == (std::vec::Vec<u8>, std::vec::Vec<u8>)`
  --> substrate/runtime-io/src/../with_std.rs:96:2
   |
96 |     triehash::trie_root(input).0
   |     ^^^^^^^^^^^^^^^^^^^ expected type parameter, found struct `std::vec::Vec`
```

I suspect it's because `Cargo.lock` lists two `triehash` crates, one from crates.io and the other from parity, both at v0.10. Could it be that parity's `triehash` should get a version bump and get pushed to crates.io?

**Comment by gavofyork** [comment]
@dvdplm this is very much inprogress - not expected to build quite yet. here just to show off direction.

**Comment by gavofyork** [comment]
that particular issue was fixed in the latest commit, fwiw; `triehash` crate had a breaking API change committed with only a tiny version increase. should really have been a minor increase.

**Comment by dvdplm** [comment]
> here just to show off direction.

Yeah, came here to look for some inspiration on generics as I struggle a bit with [this stuff](https://github.com/paritytech/parity/issues/7019).

**Comment by gavofyork** [comment]
> should be a major change if we follow semver 

it was pre-1.0, so i think minor bumps for breaking changes are the convention.

File: substrate/client/src/block_builder.rs [hunk 6]
```diff
@@ -36,7 +37,7 @@ pub struct BlockBuilder<B, E, Block, Hashing> where
 	executor: E,
 	state: B::State,
 	changes: state_machine::OverlayedChanges,
-	// will probably need PhantomData<Hashing> here...
+	dummy: PhantomData<Hashing>,
```
Comment (rphmeier) [inline substrate/client/src/block_builder.rs]: The block type ought to specify what hash algorithm it's using

**Comment by rphmeier** [comment]
Posting here so it's not hidden: The block (or even header) type ought to specify what hash algorithm it's using. Otherwise you need to specify Hashing everywhere you want to do a simple transformation from block header to hash.

**Comment by gavofyork** [comment]
Yup - should indeed be the Header so it's available within the runtime in the appropriate modules.

**Comment by rphmeier** [comment]
Ready for review. Substrate is fully generic at compile-time, but the polkadot-runtime, in order to maintain on-chain upgradeability, has to use runtime polymorphism of the extrinsic type.

`polkadot-primitives` exports an `UncheckedExtrinsic = Vec<u8>`, and `Block = generic::Block<Header, UncheckedExtrinsic`.

`polkadot-runtime` has the `Call` type available, so it exports `UncheckedExtrinsic = generic::UncheckedExtrinsic<..., Call>` and `Block = generic::Block<Header, UncheckedExtrinsic>`.

All the generic client code used in polkadot is instantiated with `polkadot_primitives::Block` as opposed to `polkadot_runtime::Block` in order to remain forwards-compatible. The only places which use `polkadot_runtime::Block` are `polkadot-consensus` (since block authorship requires knowledge of extrinsic types) and `polkadot-transaction-pool` (until we find a way to generalize meaningfully).

File: substrate/network/src/sync.rs [hunk 7]
```diff
@@ -65,16 +65,18 @@ pub enum SyncState {
...
 	/// Current global sync state.
 	pub state: SyncState,
 	/// Target sync block number.
-	pub best_seen_block: Option<BlockNumber>,
+	pub best_seen_block: Option<<B::Header as HeaderT>::Number>,
 }
 
-impl ChainSync {
+impl<B: BlockT> ChainSync<B> where
+	B::Header: HeaderT<Number=u64>,
```
Comment (gavofyork) [inline substrate/network/src/sync.rs]: Why the `u64` requirement?

File: substrate/network/src/test/mod.rs [hunk 8]
```diff
@@ -172,40 +172,38 @@ impl Peer {
...
 		if with_tx {
 			self.generate_blocks(count, |builder| {
-				let tx = test_client::runtime::Transaction {
-					from: Keyring::Alice.to_raw_public(),
-					to: Keyring::Alice.to_raw_public(),
-					amount: 1,
-					nonce: nonce,
+				let xt = Extrinsic {
+					signed: Keyring::Alice.to_raw_public().into(),
+					function: Call { to: Keyring::Alice.to_raw_public().into(), amount: 1 },
```
Comment (gavofyork) [inline substrate/network/src/test/mod.rs]: Why the change to the extrinsic format? A desire to allow misbehaviour reports in there? Can't we just make it generic and leave unimplemented (or implemented in some other way) for the test chain? I don't want to overspecialise with an assumption that each chain will always have a `Call` in there - that's kind of the point of `test_chain` having a different xt format.

File: substrate/runtime-support/src/dispatch.rs [hunk 9]
```diff
@@ -47,6 +46,9 @@ pub trait AuxCallable {
 	type Call: AuxDispatchable + Slicable + Clone + PartialEq + Eq;
 }
 
+// dirty hack to work around serde_derive issue.
```
Comment (gavofyork) [inline substrate/runtime-support/src/dispatch.rs]: Would be good to know more about this issue - a link & description of what to do when the issue is sorted.

File: substrate/runtime/primitives/src/bft.rs [hunk 10]
```diff
@@ -111,17 +110,20 @@ impl Slicable for Action {
 	}
 }
 
+/// Type alias for extracting message type from block.
+pub type MessageFor<B> = Message<B, <B as ::traits::Block>::Hash>;
```
Comment (gavofyork) [inline substrate/runtime/primitives/src/bft.rs]: What does "message" mean here? What if the block type doesn't have the notion of "message"?

File: substrate/runtime/primitives/src/generic.rs [hunk 11]
```diff
@@ -271,40 +244,76 @@ impl<Item> Slicable for Digest<Item> where
...
 	/// A chain-specific digest of data useful for light clients or referencing auxiliary data.
 	pub digest: Digest<DigestItem>,
 }
 
-impl<Number, Hash, DigestItem> Slicable for Header<Number, Hash, DigestItem> where
-	Number: Member + Slicable,
- 	Hash: Member + Slicable,
-	DigestItem: Member + Slicable,
+// Hack to work around the fact that deriving deserialize doesn't work nicely with
+// the `hashing` trait used as a parameter.
```
Comment (gavofyork) [inline substrate/runtime/primitives/src/generic.rs]: Link to issue + description of how to change once issue sorted would be good.

File: substrate/runtime/primitives/src/lib.rs [hunk 12]
```diff
@@ -55,24 +62,88 @@ pub trait BuildExternalities {
...
+impl<T> From<T> for MaybeUnsigned<T> {
+	fn from(t: T) -> Self {
+		MaybeUnsigned(t)
+	}
+}
+
+/// Verify a signature on an encoded value in a lazy manner. This can be
+/// an optimization if the signature scheme has an "unsigned" escape hash.
+pub fn verify_encoded_lazy<V: Verify, T: codec::Slicable>(sig: &V, item: &T, signer: &V::Signer) -> bool {
+	// TODO: unfortunately this is a lifetime relationship that can't
```
Comment (gavofyork) [inline substrate/runtime/primitives/src/lib.rs]: Issue link + description of how to change once issue is sorted would be good.

File: substrate/runtime/primitives/src/traits.rs [hunk 13]
```diff
@@ -148,99 +170,189 @@ impl<A: Executable, B: Executable> Executable for (A, B) {
...
-impl Header for substrate_primitives::Header {
-	type Number = substrate_primitives::block::Number;
-	type Hash = substrate_primitives::block::HeaderHash;
-	type Digest = substrate_primitives::block::Digest;
-	fn number(&self) -> &Self::Number { &self.number }
-	fn extrinsics_root(&self) -> &Self::Hash { &self.extrinsics_root }
-	fn state_root(&self) -> &Self::Hash { &self.state_root }
-	fn parent_hash(&self) -> &Self::Hash { &self.parent_hash }
-	fn digest(&self) -> &Self::Digest { &self.digest }
-	fn new(
```
Comment (gavofyork) [inline substrate/runtime/primitives/src/traits.rs]: Where did `new` go?

File: substrate/runtime/primitives/src/bft.rs (hunk 10 above)
Comment (rphmeier) [reply substrate/runtime/primitives/src/bft.rs]: Since it's in the `bft` module, it's fair to assume that this is a BFT message. The BFT subsystem is something that operates over block types as opposed to being a property of them.

File: polkadot/runtime/src/lib.rs [hunk 14]
```diff
@@ -47,28 +57,127 @@ extern crate substrate_runtime_session as session;
...
+pub type UncheckedExtrinsic = generic::UncheckedExtrinsic<AccountId, Index, Call, Signature>;
+/// Extrinsic type as expected by this runtime.
+pub type Extrinsic = generic::Extrinsic<AccountId, Index, Call>;
+
+/// Block type as expected by this runtime.
+pub type Block = generic::Block<Header, UncheckedExtrinsic>;
+
+/// Provides a type-safe wrapper around a structurally valid block.
+#[cfg(feature = "std")]
+pub struct CheckedBlock {
```
[the remark this answers is not in the document]
Comment (rphmeier) [reply polkadot/runtime/src/lib.rs]: the runtime file isn't that big, I don't see any reason to extract yet

File: substrate/network/src/blocks.rs [hunk 15]
```diff
@@ -20,47 +20,47 @@ use std::ops::Range;
...
+pub struct BlockData<B: BlockT> {
+	pub block: message::BlockData<B>,
 	pub origin: PeerId,
 }
 
 #[derive(Debug)]
-enum BlockRangeState {
+enum BlockRangeState<B: BlockT> {
 	Downloading {
-		len: BlockNumber,
```
[the remark this answers is not in the document]
Comment (rphmeier) [reply substrate/network/src/blocks.rs]: Needs to be able to convert safely between it and `usize`.

File: substrate/network/src/sync.rs (hunk 7 above)
Comment (rphmeier) [reply substrate/network/src/sync.rs]: for now, it maintains the same properties as the current sync code. it would be easy to change the sync logic to support more block number types in a follow-up PR, but since this one was getting giant I thought it better to leave that out for now.

File: substrate/runtime-support/src/dispatch.rs (hunk 9 above)
Comment (rphmeier) [reply substrate/runtime-support/src/dispatch.rs]: https://github.com/rust-lang/rust/issues/51331

see the linked serde issue

File: substrate/runtime/primitives/src/traits.rs (hunk 13 above)
Comment (rphmeier) [reply substrate/runtime/primitives/src/traits.rs]: a little above

File: substrate/runtime/primitives/src/lib.rs (hunk 12 above)
Comment (rphmeier) [reply substrate/runtime/primitives/src/lib.rs]: we would need https://github.com/rust-lang/rust/issues/44265 and then the `Fn*` traits would have to be extended to use it somehow. I don't think it's likely to be something we can work around until a few more epochs of Rust. I'll remove the `TODO` because it's not reasonable to do in the near future.
